### PR TITLE
[WIP] Add grouped_matmul_swiglu_quant_v2_layered CAM operator

### PR DESCRIPTION
**Pull request by simonkorl**
## Purpose

Vendor the grouped_matmul_swiglu_quant_v2_layered operator (and its shared headers) from the openEuler umdk `cam` `comm_operator` tree (`cam_feature` @ 0281a98a) into `csrc/npu/ascend_kernels/`, so AFD can build it from source alongside the existing a2e/e2a operators.

## Issue

- Related issue(s): #364

## Scope

- In scope: grouped_matmul_swiglu_quant_v2_layered op dir + shared
  tiling_base/err/ops_*/layered headers; operator_registry.json entry;
  op_host include-path port; pybind entry point.
- Out of scope: other cam_feature operators; connector wiring; quantization
  correctness benchmark (tracked separately).

## Implementation Notes

- Verbatim vendor (MIT/Huawei headers preserved, dual CANN v2.0 + MIT header
  on the kernel source noted); verified SHA-256 identical to umdk cam_feature.
- No vLLM source changes; plugin-owned custom-op build path only.

## Test Plan

- CPU-safe: registry/select_ops unit checks; pybind import guard.
- Ascend-gated: csrc/npu run-package build and CAM async DeepSeek-V3.2 E2E.

## Test Result

image

## Docs Impact

- Files updated: csrc/npu/README.md

---

<details>
<summary>Essential PR Checklist</summary>

- [x] Purpose is clear and linked to public context when possible.
- [x] Scope is bounded.
- [x] Compatibility with vLLM v0.26.0 is considered.
- [x] No changes are made to the vLLM source checkout.
- [x] Plugin-owned classes or explicit dotted class paths are preferred over monkey patches.
- [x] Any compat shim or monkey patch is isolated, idempotent, version-guarded, documented, and tested.
- [x] Imports remain CPU-safe; CUDA-heavy work is delayed or GPU-gated.
- [x] Validation evidence is included, including skipped GPU tests when applicable.
- [x] Documentation impact is stated.

</details>

### CODE REVIEW

File: csrc/npu/ascend_kernels/grouped_matmul_swiglu_quant_v2_layered/op_host/op_tiling/grouped_matmul_swiglu_quant_v2_layered_tiling.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    compileInfoPtr->aivNum_ = ascendcPlatform.GetCoreNumAiv();
+    std::string platformRes;
+    platformInfoPtr->GetPlatformRes("AICoreintrinsicDtypeMap", "Intrinsic_data_move_l12bt", platformRes);
+    compileInfoPtr->supportL12BtBf16 = (platformRes.find("bf16") != std::string::npos);
+    ascendcPlatform.GetCoreMemSize(platform_ascendc::CoreMemType::UB, compileInfoPtr->ubSize_);
+    OP_LOGD(context->GetNodeName(), "ubSize is %lu, aicNum is %u.", compileInfoPtr->ubSize_, compileInfoPtr->aicNum_);
+    return GRAPH_SUCCESS;
+}
+
+IMPL_OP_OPTILING(GroupedMatmulSwigluQuantV2Layered)
```
Comment (jiangkuaixue123) [inline csrc/npu/ascend_kernels/grouped_matmul_swiglu_quant_v2_layered/op_host/op_tiling/grouped_matmul_swiglu_quant_v2_layered_tiling.cpp:84]: [P1] Include the nested tiling sources in the host build

The staging script preserves this file and the base/fusion tiling implementations under op_host/op_tiling/, but cmake_files/op_host/CMakeLists.txt still uses aux_source_directory(${CMAKE_CURRENT_SOURCE_DIR} ops_srcs), which only collects the top-level sources. Neither code generation nor cust_optiling receives these three .cpp files, so this operator's tiling implementation and registration are omitted from the source-built package. Explicitly add the nested tiling sources to the appropriate build inputs and verify that the resulting library contains the registration; the registry/Meta tests cannot detect this omission.

File: csrc/npu/ascend_kernels/grouped_matmul_swiglu_quant_v2_layered/op_api/gmm_dsq_base.h [hunk 2]
```diff
@@ -0,0 +1,636 @@
...
+        OP_CHECK_WRONG_DIMENSION(gmmDsqParams_.groupList, GROUP_LIST_DIM_LIMIT, return false);
+        OP_CHECK_WRONG_DIMENSION(gmmDsqParams_.output, QUANTOUT_DIM_LIMIT, return false);
+        OP_CHECK_WRONG_DIMENSION(gmmDsqParams_.outputScale, QUANTSCALEOUT_DIM_LIMIT, return false);
+        return true;
+    }
+
+    bool CheckInputOutDimsA4W4orA8W4()
+    {
+        OP_CHECK_WRONG_DIMENSION(gmmDsqParams_.x, X_DIM_LIMIT, return false);
+        if (gmmDsqParams_.isA4W4 && gmmDsqParams_.weightAssistMatrix != nullptr) {
```
Comment (jiangkuaixue123) [inline csrc/npu/ascend_kernels/grouped_matmul_swiglu_quant_v2_layered/op_api/gmm_dsq_base.h:100]: [P1] Normalize an absent assist matrix before A4W4 validation

The Python schema requires Tensor[] for all_weight_assist_matrix, so callers represent an absent matrix with []. ConvertType(TensorList) passes that through aclCreateTensorList rather than explicitly converting it to nullptr. This check rejects a non-null empty-list handle on the A4W4 path. CheckOptionalTensorListEmpty() only runs later in CovertDataContiguous(), after Process() has already called CheckParams(), so it cannot make this input pass validation. Normalize the optional list before CheckParams(); also make sure the later empty placeholder is compatible with the tiling list-length check when more than one layer is supplied. Add a real A4W4 invocation test with an empty assist list, since Meta dispatch bypasses this validation.

**Review by jiangkuaixue123 — COMMENTED** [review]
Static review of adf43d2: two issues in the source-build and A4W4 invocation paths. Ascend build and numerical execution were not run; the added Meta tests do not exercise these paths.
